Sources/Workspace/LocalContainerProvider.swift

```diff
@@ -0,0 +1,111 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import PackageModel
+import PackageGraph
+import TSCBasic
+import TSCUtility
+import SourceControl
+
+/// PackageContainerProvider implementation used by Workspace to do a dependency pre-calculation using the cached
+/// dependency information (Workspace.DependencyManifests) to check if dependency resolution is required before
+/// performing a full resolution.
+struct LocalContainerProvider: PackageContainerProvider {
+    let root: PackageGraphRoot
+    let currentManifests: Workspace.DependencyManifests
+    let config: SwiftPMConfig
+    let diagnostics: DiagnosticsEngine
+
+    func getContainer(
+        for identifier: PackageReference,
+        skipUpdate: Bool,
+        completion: @escaping (Result<PackageContainer, AnyError>) -> Void
+    ) {
+        if let manifest = currentManifests.dependencies.first(where: { $1.packageRef == identifier }) {
+            let container = LocalPackageContainer(
+                manifest: manifest.manifest,
+                dependency: manifest.dependency,
+                config: config
+            )
+
+            return completion(.success(container))
+        }
+
+        if let index = currentManifests.root.packageRefs.firstIndex(of: identifier) {
+            let container = LocalPackageContainer(
+                manifest: currentManifests.root.manifests[index],
+                dependency: nil,
+                config: config
+            )
+
+            return completion(.success(container))
+        }
+
+        completion(.init(Diagnostics.fatalError))
+    }
+}
+
+private struct LocalPackageContainer: PackageContainer {
+    let manifest: Manifest
+    let dependency: ManagedDependency?
+    let config: SwiftPMConfig
+
+    var identifier: PackageReference {
+        if let identifier = dependency?.packageRef {
+            return identifier
+        } else {
+            let identity = PackageReference.computeIdentity(packageURL: manifest.url)
+            return PackageReference(
+                identity: identity,
+                path: manifest.path.pathString,
+                isLocal: true
+            )
+        }
+    }
+
+    var reversedVersions: [Version] {
+        guard let dependency = self.dependency else { return [] }
+
+        switch dependency.state {
+        case .checkout(let state):
+            if let version = state.version {
+                return [version]
+            }
+        case .edited, .local:
+            break
+        }
+
+        return []
+    }
+
+    func isToolsVersionCompatible(at version: Version) -> Bool {
+        return true
```

Comment:
should we add an assert here?

Reply:
I did a check against the current tools version.